### PR TITLE
fix write_minimizer

### PR DESCRIPTION
**Pull request by lrobidou**
# Description

A part of the `write_minimizer` function seems odd to me. Using the version I propose solves an issue I have when using this library. I am unsure if I understand the code of this crate well enough, please review this change with care.

If I understand, the idea is to skip the bases from `(self.minimizer_offset * 2)` to `(self.minimizer_offset * 2) + m * 2`, which I don't think is what the code did.

The fix does not break any tests, but I guess it's because the tests do not cover this part of the code.

## Type of change

Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I run `cargo fmt`
- [x] I run `cargo clippy`
- [x] I run `cargo test`
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes
- [ ] I have check my test covered almost all new code

### CODE REVIEW

**Comment by lrobidou** [comment]
I also need a way to extract kmers from each minimizer block individually.
Something like:
````
let file = kff::Kff::<std::io::BufReader<std::fs::File>>::open(kff_path).unwrap();

for block in file.minimizer_blocks() {  // this
    for kmer in block.kmers() {
        // do something with the kmer
    }
    // end of processing of a block
}
````
Is it possible to do this with this current crate?

**Comment by natir** [comment]
> The fix does not break any tests, but I guess it's because the tests do not cover this part of the code.
Tarpaulin say this code is covered but I can reproduce your statement it's strange.

Can you provides a minimal example of data that produce bug you talk about or add a test in your PR for this bug.

About you request to iterate block by block you can use [next_kmer_section](https://docs.rs/kff/latest/kff/kff/struct.Kff.html#method.next_kmer_section) you could see an [*example*](https://github.com/Kmer-File-Format/kff-rust-api/blob/main/src/iterator.rs#L52) in iterator.

**Comment by lrobidou** [comment]
I added a unit test that fails without the patch (at least on my machine...).

**Comment by lrobidou** [comment]
I also added some documentation on the README and made public the encoding functions. Don't know if that's OK?

**Comment by lrobidou** [comment]
There are still some TODOs in the README. I'm not sure I'm able to complete the README as I did not use these sections.

**Comment by lrobidou** [comment]
Just to be sure, I re-cloned the repo from your main branch, added the test and it failed.
I'm running a debian distribution. My CPU is a 13th Gen Intel (i7-1365U), x86_64, Little Endian.

**Comment by lrobidou** [comment]
I usually prefer projects having an overview in the README as it prevents me from clicking on multiple links (I'm lazy). Plus it's easy to do a PR  to correct it.
But I have no issue with putting the doc online if you prefer. Add somewhere on the doc how to modify it if it is possible.
It would be nice if you could put a link to my GitHub account on the doc somewhere (so people know who to complain about when this does not compile). Or maybe I can do a PR on the doc directly?

**Comment by natir** [comment]
On my way I didn't like large Readme.

Just by edit file in [website directory](https://github.com/Kmer-File-Format/kff-rust-api/tree/main/website) github action rebuild and publish manual at each commit on main branch.

I would be very grateful if you would create a PR to update the manual. You can also add links in the readme to point directly to the manual, but also create a contributor section and add your name.

**Comment by natir** [comment]
I reproduce the bug !

**Comment by lrobidou** [comment]
I will do this tomorrow, I'm on UTC+9 for a few weeks.

**Comment by natir** [comment]
Thank a lot for your help !

**Comment by natir** [comment]
I'm sorry it took me a while to figure out the bug,  I'm having a hard time getting back into the code, thanks for your help.

It's all `+ 1`'s fault, the resize has no impact on the tests, it's there to reassure me that the minimizer is the right size. I'm not sure I've written a test case that makes it necessary, but I'm more reassured with it.

I'll merge your Pull Request as soon as you've moved the documentation to another PR.

Thanks again for your work!

**Comment by lrobidou** [comment]
I think it's also because the `* 2` applied to `m` and not to `self.minimizer_offset`.
( self.minimizer_offset + m * 2 ) + 1
vs
( self.minimizer_offset + m ) * 2

**Comment by lrobidou** [comment]
Regarding the resize, I would be more inclined to add some `debug_assert!` (I could also use `assert!`, as these checks are there to protect from undefined behavior). They are removed when compiling in release mode, so they have no impact on the performance. And they crash instead of failing silently, which I think is better.


I can open another PR and add some assertions to check inputs coming from the user (knowing I used the lib badly would have helped me, so I want to help others).

**Comment by lrobidou** [comment]
By the way, I only added myself in the contributor section. I did not know if you wanted to appear in another section (`Maintainers` ?) and/or which name to use. Remember to add yourself after the merge :)

**Comment by lrobidou** [comment]
Also, you may squash my commits to keep your history clean, I have no preference over this.

**Review by natir — APPROVED** [review]

**Pull request merged**
